[idlharness.js] Simplify handling of inheritance and mixins

foolip opened the pull request:
Handling of these concepts was more complicated than (now) necessary.

No changes in the actual test results are intended.

Inheritance:

There's no reason to record inheritance separately in `IdlArray`'s
`this.["inheritance"]`, since it can be determined just as easily
from `this.members` via the `member.base` attributes.

The concept of "consequential interfaces" in Web IDL went away with
`implements` statements in https://github.com/heycam/webidl/pull/433
and here in https://github.com/web-platform-tests/wpt/pull/28619.
`traverse_inherited_and_consequential_interfaces()` can be replaced
with just `get_inheritance_stack()`.

While this changes the behavior of `default_to_json_operation()`, there
is no `toJSON` operation declared in a mixin in interfaces/, only in
resources/ tests, so this change should not affect real tests.

default_to_json_operation.html was updated along the lines of simplified
examples in the spec: https://github.com/heycam/webidl/pull/980

Mixins:

For valid `A includes B` statements, `A` is always an interface and
`B` is always an interface mixin, so there are no include chains or
the possibility of cycles. `recursively_get_includes()` assumed this.

Instead just save the `includes` statements found in `this.includes`
and apply them in `merge_mixins`, similar to partials.

The handling of partials isn't changed, but `collapse_partials` is
renamed to `merge_partials` to match the above.

Not today:

Further unification of partials and mixins is possible, and the handling
of [Exposed] is missing for mixins, but this is left for later.

foolip commented:
@Ms2ger I expect tests will still fail here, but initial review would still be great.

Ms2ger commented on resources/idlharness.js:
```diff
@@ -177,27 +177,9 @@ self.IdlArray = function()
      * on a later one.  Save these up and handle them right before we run
      * tests.
-     *
-     * .partials is simply an array of objects from WebIDLParser.js'
```
Let's keep this comment about `partials` and add one about `includes`

foolip replied:
OK, I'll rewrite it slightly but keep the gist of it.

foolip committed:
[idlharness.js] Simplify handling of inheritance and mixins

Handling of these concepts was more complicated than (now) necessary.

No changes in the actual test results are intended.

Inheritance:

There's no reason to record inheritance separately in `IdlArray`'s
`this.["inheritance"]`, since it can be determined just as easily
from `this.members` via the `member.base` attributes.

The concept of "consequential interfaces" in Web IDL went away with
`implements` statements in https://github.com/heycam/webidl/pull/433
and here in https://github.com/web-platform-tests/wpt/pull/28619.
`traverse_inherited_and_consequential_interfaces()` can be replaced
with just `get_inheritance_stack()`.

While this changes the behavior of `default_to_json_operation()`, there
is no `toJSON` operation declared in a mixin in interfaces/, only in
resources/ tests, so this change should not affect real tests.

default_to_json_operation.html was updated along the lines of simplified
examples in the spec: https://github.com/heycam/webidl/pull/980

Mixins:

For valid `A includes B` statements, `A` is always an interface and
`B` is always an interface mixin, so there are no include chains or
the possibility of cycles. `recursively_get_includes()` assumed this.

Instead just save the `includes` statements found in `this.includes`
and apply them in `merge_mixins`, similar to partials.

The handling of partials isn't changed, but `collapse_partials` is
renamed to `merge_partials` to match the above.

Not today:

Further unification of partials and mixins is possible, and the handling
of [Exposed] is missing for mixins, but this is left for later.

foolip commented on resources/idlharness.js:
```diff
@@ -178,26 +178,11 @@ self.IdlArray = function()
-     *   interface B : C {};
-     *
-     * results in this["inheritance"] = { A: "B", B: "C" }
+     * Both this.partials and this.includes will be the objects as parsed by
```
As hinted at in the commit message, it would be natural to add `this.mixins` and not wrap mixins in `IdlInterface` at all. But not yet.

foolip commented on resources/idlharness.js:
```diff
@@ -249,11 +234,15 @@ IdlArray.prototype.internal_add_dependency_idls = function(parsed_idls, options)
 
     const all_deps = new Set();
-    Object.values(this.inheritance).forEach(v => all_deps.add(v));
-    // NOTE: If 'A includes B' for B that we care about, then A is also a dep.
```
This comment was misleading, there is no "that we care about" condition here.

foolip commented on resources/test/tests/functional/idlharness/IdlInterface/test_partial_interface_of.html:
```diff
@@ -80,17 +80,20 @@
+        assert_equals(includes.get('J'), 'G', 'J should be picked up');
+        // K isn't a dep because J isn't defined.
         assert_false('K' in idlArray.members, 'K should be ignored');
-        // L inherits F, so should be picked up.
```
This comment and the assert description was wrong, now fixed.

foolip commented on resources/test/tests/unit/IdlInterface/default_to_json_operation.html:
```diff
@@ -128,45 +90,24 @@
         assert_array_equals([...map.values()].map(v => v.idlType), ["DOMString", "long"]);
     }, 'should return a properly ordered map that accounts for mixed-in interfaces which declare a [Default] toJSON operation.');
-
-    test(function() {
```
These tests are just more convoluted forms of the earlier tests IMHO. Order, the non-inheritance of `toJSON` and the (now correct) mixin behavior is tested above.

foolip commented:
@Ms2ger fixing the resources/ tests revealed that I had changed the order (fixed by add `.reverse()`) and I went down a rabbit hole with https://github.com/heycam/webidl/issues/979, but now I think it's all in order. It could still be simplified more, but not all in one PR.

Ms2ger commented on resources/idlharness.js:
```diff
@@ -1414,9 +1365,9 @@ IdlInterface.prototype.get_inheritance_stack = function() {
+IdlInterface.prototype.default_to_json_operation = function() {
     var map = new Map(), isDefault = false;
-    this.traverse_inherited_and_consequential_interfaces(function(I) {
+    this.get_inheritance_stack().reverse().forEach(function(I) {
```
This made me a little nervous because `reverse` operates in-place, and `get_inheritance_stack` doesn't signal very clearly that it returns a fresh array. It seems all is fine, though.

Followup: of the three callers of `get_inheritance_stack`, one only calls it for the side effects and the two others iterate backward. Maybe change `get_inheritance_stack` to return the array reversed already?

foolip replied:
I'll send a follow-up to rename it to `get_reverse_inheritance_stack`.

foolip replied:
https://github.com/web-platform-tests/wpt/pull/28688

Ms2ger reviewed (approved):
Haven't done a particularly deep review, but this all seems fine.
